underscore.js

```diff
@@ -490,17 +490,21 @@
   };
 
   // Internal implementation of a recursive `flatten` function.
-  var flatten = function(input, shallow, strict) {
-    var output = [], idx = 0;
+  var flatten = function(input, shallow, strict, _output, _idx) {
+    var output = _output ? _output : [], idx = _idx ? _idx : 0;
     for (var i = 0, length = getLength(input); i < length; i++) {
       var value = input[i];
       if (isArrayLike(value) && (_.isArray(value) || _.isArguments(value))) {
         //flatten current level of array or arguments object
-        if (!shallow) value = flatten(value, shallow, strict);
-        var j = 0, len = value.length;
-        output.length += len;
-        while (j < len) {
-          output[idx++] = value[j++];
+        if (!shallow) {
```

Comment:
Do you mind swapping the two conditional paths? Reading a `!condition` with an `else` always makes my head hurt.

```patch
if (shallow) {
  var j = 0, len = value.length
  //...
} else {
  flatten(value, shallow, strict, output, idx);
  idx = output.length;
}
```